pkg/controller/controller.go

```diff
@@ -326,6 +356,48 @@ func NewCSIProvisioner(client kubernetes.Interface,
 		extraCreateMetadata:                   extraCreateMetadata,
 		eventRecorder:                         eventRecorder,
 	}
+	if nodeDeployment != nil {
+		// The base delay configures the rate limiting which
+		// prevents concurrent external-provisioner instances
+		// to start working on a new PVC all at the
+		// sametime. This delay also influences the jitter, so the
+		// initial delay will be somewhere between 0 and
+		// baseDelay.
+		//
+		// It must be large enough that one instance has the chance to
+		// become owner of a PVC and get that change propagated to the
+		// other instances before those wake up to try that
+		// themselves.
+		//
+		// With a value of 10 seconds, when creating 5000
+		// volumes on a cluster with 50 instances only ~300
+		// update conflicts occurred. With a value of 1 second,
+		// over 6000 conflicts occurred. The rate of volume
+		// provisioning was the same in both cases.
+		//
+		// The latency per volume was probably higher with the
+		// higher delay, but that was not measured.
+		//
+		// It might make sense to make this value configurable so
+		// that CSI driver deployments can tweak it depending
+		// on their needs.
+		baseDelay := 10 * time.Second
```

Comment:
This IMO needs to be configurable or grow with the cluster size (as a separate PR)

Reply:
This got me thinking. The current implementation does exponential backoff *per PVC*. This is almost never going to make a difference in practice because after the initial conflict, the PVC has a selected node and there's no need to do another update attempt for it. The only exception is when provisioning on that selected node fails and there has to be a global retry - that should be rare.

In this approach, we can't make this grow with cluster size either: we cannot assume that the CSI driver runs on all nodes. It might manage specialized resources that are only available on some nodes. With deployment handled by Kubernetes, the individual instances have no information which other instances exist.

I had considered making this independent of the PVC. The issue with that is the question of when to reduce the delay and how much. When the "winning" instance resets it to zero, then it is also likely to win for the next PVC until its space is exhausted and it stops trying to own a PVC (because of the capacity check). I preferred even spreading across the cluster. But perhaps filling up one node is also fine: if users want a certain policy for volume scheduling, they should use late binding and influence volume placement indirectly through that.

 I think I will give that a try...

Reply:
It ended up working worse. The main problem was that the instance with the smallest backoff delay ends up grabbing more volumes for provisioning than what it has space for, so once it is the owner it fails for at least some of them. Those then are made available to other instances, but the same pattern just repeats.

It could be made to work if the provisioner could model remaining capacity considering how many volumes already were assigned, but that is driver specific and complex.

This is more of a problem under load (= many pending PVCs) and a full cluster (= capacity exhausted on some nodes), but still, when volumes are spread evenly purely based on statistics, it worked better.

So I'll stick with the current approach and perhaps tweak it a bit: when tracking the number of collisions per number of PVCs, it should be possible to scale the base delay.

Reply:
> So I'll stick with the current approach and perhaps tweak it a bit: when tracking the number of collisions per number of PVCs, it should be possible to scale the base delay.

That has the same problem. When using a running average of the failure rate as indicator for "delay is too low", I saw increasing numbers of failed provisioning attempts because of full nodes.

TLDR: cluster admins will have to adjust the delay for the expected cluster size and number of volumes.